AE2 conduits crashes server on chunkload

jollycar opened the pull request:
# Description

If you try to create a IManagedGridNode inside an unloaded chunk, AE2 will throw an IllegalStateException.

The documentation for IManagedGridNode.create() is clear: "This should only be called when the node is in a ticking chunk". The suggested solution is GridHelper.onFirstTick().

The fix: When calling AE2InWorldConduitNodeHost.onCreated(), use 
GridHelper.onFirstTick(). Also implement the review suggestions. 

Closes #434

jollycar committed:
AE2 conduits crashes server on chunkload

If you try to create a IManagedGridNode inside an unloaded chunk,
AE2 will throw an IllegalStateException.

The documentation for IManagedGridNode.create() is clear:
"This should only be called when the node is in a ticking chunk".
The suggested solution is GridHelper.onFirstTick().

The fix: When calling ConduitBlockEntity.onload(), make sure at least
1 server tick has happened by calling checkConnection.isInitialized().

closes #434

justliliandev commented:
this will not work when ae2 is installed as then GridHelper can't be found

jollycar commented:
I think you mean "when ae2 is NOT installed". If so, I understand your answer and will try to work around this.

justliliandev commented:
yeah, meant that, sorry, I'm in a hurry rn

justliliandev commented:
scheduling that inside AE2InWorldConduitNodeHost#onCreated solves the classloading problem of this, not sure if this really works as ae2 documentation was out of data before

jollycar commented:
New fix doesn't involve GridHelper and works on the world privided by Casimir255

Edit: Had to add a another commit, because I had not removed the imports. (I blame IntelliJ)

jollycar committed:
Oops remove imports

jollycar committed:
Merge branch 'Team-EnderIO:dev/1.20.1' into fix/ae2-conduid-crash

justliliandev reviewed (changes requested):
this might work, but:
1. check connections isn't the right thing to use, it's for scheduling a recheck for connections.
2. this breaks node recovery, as the code now in your added check never runs
3. onCreated is called twice for lazynodes (that's an issue made by me)
Solutions for this:
1. move the bundle#onLoad call to before the lazyNode loop
2. remove data#onCreated from the lazyNode loop
3. Use GridHelper inside the AE2InWorldConduitNodeHost#onCreated to delay it's initialization
If the world doesn't break with all 3 things applied, then this is probably the right solution for the problem.

Also: I'm sorry for this mess and thanks for looking into this

jollycar committed:
Merge branch 'Team-EnderIO:dev/1.20.1' into fix/ae2-conduid-crash

jollycar committed:
AE2 conduits crashes server on chunkload

If you try to create a IManagedGridNode inside an unloaded chunk,
AE2 will throw an IllegalStateException.

The documentation for IManagedGridNode.create() is clear:
"This should only be called when the node is in a ticking chunk".
The suggested solution is GridHelper.onFirstTick().

The fix: When calling AE2InWorldConduitNodeHost.onCreated(), use
GridHelper.onFirstTick(). Also implement the review suggestions.

closes #434

jollycar commented:
> this might work, but:
> 
>     1. check connections isn't the right thing to use, it's for scheduling a recheck for connections.
> 
>     2. this breaks node recovery, as the code now in your added check never runs
> 
>     3. onCreated is called twice for lazynodes (that's an issue made by me)
>        Solutions for this:
> 
>     4. move the bundle#onLoad call to before the lazyNode loop
> 
>     5. remove data#onCreated from the lazyNode loop
> 
>     6. Use GridHelper inside the AE2InWorldConduitNodeHost#onCreated to delay it's initialization
>        If the world doesn't break with all 3 things applied, then this is probably the right solution for the problem.
> 
> 
> Also: I'm sorry for this mess and thanks for looking into this

Implemented the changes. The world save from Casimir255 also seems to work with these changes.
